### ISSUE TITLE
MPCS 51215 Commentary, HW1

### ISSUE THREAD

**Issue by cfernanne** [dsahota-applied-data-analysis/project-report-cfernanne#2]
**Purpose:**
This code aims to construct a classifier which identifies political bias in newsmedia text via NLP. I consider a repository of news articles from the last odd decade, pre-classified by experts at AllSides, and built out a transform model (distiliBERT) which identifies left, center, or right lean (L, C, R) in a given news headline. We compare this to a baseline TF-IDF (modified bag of words) + logistic regression model, hoping for substantial improvement. Such a classifier could be useful to editors and newsrooms attempting to flag their own bias, rather than merely consumers, for whom resources that gauge bias focus on biased news *sources*, rather than the text itself.

**Instructions on how to run:**
The repo features three separate ipynb files, which each build out various classifiers after breaking down the dataset differently:
- Final_Project (Tripartite).ipynb: trains and fits five 3-way classifiers on the full dataset, computes and displays SHAP values for transform model ~ 400 lines of code
- Final_Project (Binary).ipynb: trains and fits five binary classifiers on the full dataset, ~200 lines of code
- Final_Project (Binary) - Medicine Specific.ipynb: trains and fits five binary classifiers on only the articles topical to healthcare, ~200 lines of code
The first contains the final model, which is trained and tested under the "Fine-Tuned DistilBert Model" header. This is followed by a SHAP analysis. This can be run locally via Jupyter Notebook (git clone -> cd project-report-cfernanne -> pip install notebook -> jupyter notebook, after which a browser tab will open automatically where you can click on the .ipynb file and run it cell by cell).

**Does this code solve the problems you wanted it to solve?:**
Not quite. The three-way model I landed on is of poor general accuracy, struggling to differentiate between center and left leaning text, but does effectively flag pieces with a right lean and outperform a baseline model (TF-IDF + logistic regression). The overall accuracy of using a distilBERT model was underwhelming, not a large improvement over a simpler BoW model, both in the three-way and modified, two-way classification problem (unbiased--C or biased--L || R). This is likely because I’m imposing limits on batch size and epochs based on time and computation power. 

More specifically, we see shifts in class-wise performance; the baseline model showed more balanced recall across all three classes (left: 36%, center: 45%, right: 32%), whereas DistilBERT tends to predict "right" bias (recall = 73%) but struggled with "left" and "center" (both at 23%). The dataset was balanced between the three categories, so this suggests that the model is able to flag R relatively easily, but struggles to differentiate between C and L. This intuitively follows from the state of modern print newsmedia, which is largely center-left (whereas conservative media tends to be more bold-facedly right wing). 

**What are you proud of about this code?:** 
This was my first time delving into NLP, so I'm pleased that I was generally able to be self-directed throughout a particularly applied introduction to the concept as I built out a project from start to finish. I think the different methods I read about for featurizing text were interesting, from simple approaches like the intuitive BoW, to TF-IDF (which complicates things slightly by prioritizing important words over frequent but uninformative words), to transformer-based models with a lot of wacky things going on under the hood. 

I think it’s also interesting how you can pick a model’s brain, so to speak, with SHAP and visualize feature importance in a way that remains intelligible, if only as a sense check. I learned about violin and dot summary plots of SHAP values in the process of putting the code together (see bottom portion of Final_Project (Tripartite).ipynb, particular shap.summary_plot()s ) and was proud to see the resultant figures did sense-check, suggesting that the model is picking up on some signal rather than noise/confounders (e.g. features of a particular paper's style guide or verbiage rather than *content* that is politically biased).

**Is there anything about this code that embarrasses you?:**
The accuracy is very poor for the final three-way model, scarcely better than random chance:

              precision    recall  f1-score   support

        left       0.44      0.23      0.30       850
      center       0.47      0.23      0.31       851
       right       0.37      0.73      0.49       851

    accuracy                           0.40      2552
   macro avg       0.42      0.40      0.37      2552
weighted avg       0.42      0.40      0.37      2552

I'm concerned this is in part because I featurized the data incorrectly, maybe by running the headlines through the standard spaCy features before applying the tokenizer from DistilBertTokenizerFast (see "processed_heading," rather than "headings" in the tokenize_function()), which may have filtered out some signal that the transform model could have better leveraged. If so, it was a silly mistake and would only be remedied by rerunning all the transform training, tuning, and testing all over again, which is very computationally intensive. 

**Is there anything about this code that you don’t understand? A bug, for example, that you’re not sure how to fix?** 
Not a bug, but simply getting worse accuracy than I would anticipate, as above. I'm unsure as to how to strike a balance between computational efficiency (particularly in the training and tuning stages) and accuracy (which is, of course lost when we restrict our dataset to just news headlines rather than entire excerpts). Final_Project (Binary).ipynb and Final_Project (Binary) - Medicine Specific.ipynb sees me attempt to simplify to a binary classifier and narrow down the types of news articles to standardize the jargon, but this didn't seem to have a substantial effect on accuracy. I just don't know what variable I should tune/how to best constrain my question without unworkably slow training times.

**Do you have any other unanswered questions that came up for you while you were developing this code?** 
Though this was partially assuaged by the (very limited) SHAP analysis, I'm concerned about the classifier reverse-engineering media *source* rather than drawing conclusions from the text; that is, instead of "Oh, this article uses X language -> it must be R," the classifier is going "Oh, this article uses Y language -> which is consistently how Fox news structures headlines -> Fox news is R -> the article must be R". This is undesirable, because the classifier should theoretically be usable in a newsroom and so classify purely based on bias in the text. I could've tested this by holding out some newssources from the training data and checking if testing on the holdouts is substantially worse (suggesting that high performance boils down to memorizing source conventions). There also may be debiasing techniques to correct this, but they are presently beyond my ken.
